# Day Four: for loops
Tutorial
In this challenge, you'll need to use a for loop to reach the answer. In programming, a loop is a block of code that's executed repeatedly, either for a set number of times, or until a certain condition is met. A for loop is a loop that is executed for a set number of times, iterated over the number of items in a sequence. To learn more about for loops, read this article.

lst = [1,2,3,4,5]


for i in range(len(lst)): 
    '''insert code'''
In the above code, the range() function creates a sequence of numbers, and the len() function outputs the length of a list.

Challenge:
Dot needs to purchase:

- 600 planks of Oak Wood
- 150 liters of Blue Paint
- 15 liters of White Paint
- 165 liters of Paint Finish


- Item	Needed Amount to Buy	Wholesale Price	Retail Price
- Plank of Oak Wood	600	$ 7000	$ 12.99
- 1 Liter of Blue Paint	150	$ 1000	$ 8.99
- 1 Liter of White Paint	15	$ 1000	$ 9.99
- 1 Liter of Paint Finish	165	$ 800	$ 3.99

Use a loop to determine the price Dot would pay for purchasing supplies at the retail price. Based on that calculation, which itmes should Dot buy at retail vs. wholesale?

Note: Assume the wholesale price covers all the supply Dot needs for each item, whereas the retail price is per single unit.


In [1]:
item_list = ['Oak Wood', 'Blue Paint', 'White Paint', 'Paint Finish']

amount_list = [600,150,15,165]

wholesale_price_list = [7000, 1000, 1000, 800]

retail_price = [12.99, 8.99, 9.99, 3.99]

In [2]:
for i in range(len(item_list)):
    print(item_list[i])

Oak Wood
Blue Paint
White Paint
Paint Finish


In [4]:
for i in range(len(item_list)):
    print(item_list[i],wholesale_price_list[i], amount_list[i]*retail_price[i])

Oak Wood 7000 7794.0
Blue Paint 1000 1348.5
White Paint 1000 149.85
Paint Finish 800 658.35


More easy to do it like:

In [9]:
full_price = []
for i in range(len(item_list)):
    
    total_price = amount_list[i] * retail_price[i]
    full_price.append(total_price)

In [10]:
print(full_price)

[7794.0, 1348.5, 149.85, 658.35]


# Day 5: conditional logic and control flows
Conditional logic is essential to control flow statements. In programming, control flows are expressed in if...else statments. To read more on if...else statements, read this article before proceeding.


#If Statment
if condition is True:
    '''
    execute body of code

'''

#If Else Statments
if var > 10: #if var is greater than 10
    '''
    execute body of code
    '''
else: #if none of the above statement is true, execute code
    '''
    execute body of code
    '''


#Mutliple If...else statements
if var > 10: #if var is greater than 10
    '''
    execute body of code
    '''

elif var > 5 and <= 10: #if var is greater than 5 but less than or equal to 10
    '''
    execute body of code
    '''
else: #if none of the above statement is true, execute code
    '''
    execute body of code
    '''
    
 Challenge
Dot's okay with paying a bit of a surplus for convenience, but they don't want to go broke buying dustpans and glass cleaner. Help them figure out which items cost over 10% more at the nearby store, so they can avoid buying these items.

Using Python, develop a list of items that are too expensive for Dot to purchase at the nearby store.

In [11]:
# Cleaning Supplies List (19 items)
cleaningsupplies_list = ['Broom', 'Mop', 'Dustpan', 'Garbage Bags', 'Glass Cleaner', 'Vinegar',
                        'Soap', 'Bleach', 'Duster', 'Floor Cleaner', 'Sponges', 'Dish Soap',
                        'Drain Cleaner', 'Paper Towels', 'Cleaning Rags', 'Toilet Cleaner', 
                        'Rubber Gloves', 'Alcohol Wipes', 'Squeegee']

# City Price
city_price = [6.49, 4.99, 3.39, 4.29, 3.99, 1.99, 
              1.50, 3.99, 4.99, 5.99, 2.99, 2.99, 
              5.99, 2.99, 3.49, 6.99, 2.99, 1.98, 11.99]

# Country Price
country_price = [5.49, 4.69, 4.42, 5.99, 5.99, 2.50,
                1.25, 2.49, 4.50, 6.75, 2.49, 1.99, 
                6.25, 3.99, 3.59, 4.99, 1.69, 1.87, 10.99]

In [13]:
for i in range(len(cleaningsupplies_list)):
    print(cleaningsupplies_list[i])

Broom
Mop
Dustpan
Garbage Bags
Glass Cleaner
Vinegar
Soap
Bleach
Duster
Floor Cleaner
Sponges
Dish Soap
Drain Cleaner
Paper Towels
Cleaning Rags
Toilet Cleaner
Rubber Gloves
Alcohol Wipes
Squeegee


In [14]:
for i in range(len(cleaningsupplies_list)):
    print(cleaningsupplies_list[i], country_price[i]/city_price[i])

Broom 0.8459167950693375
Mop 0.9398797595190381
Dustpan 1.303834808259587
Garbage Bags 1.3962703962703964
Glass Cleaner 1.5012531328320802
Vinegar 1.256281407035176
Soap 0.8333333333333334
Bleach 0.6240601503759399
Duster 0.9018036072144289
Floor Cleaner 1.1268781302170283
Sponges 0.8327759197324415
Dish Soap 0.6655518394648829
Drain Cleaner 1.0434056761268782
Paper Towels 1.334448160535117
Cleaning Rags 1.0286532951289398
Toilet Cleaner 0.7138769670958512
Rubber Gloves 0.5652173913043478
Alcohol Wipes 0.9444444444444445
Squeegee 0.9165971643035863


In [16]:
for i in range(len(cleaningsupplies_list)):
    if  country_price[i]/city_price[i] >= 1.1:
        print(cleaningsupplies_list[i], country_price[i]/city_price[i])

Dustpan 1.303834808259587
Garbage Bags 1.3962703962703964
Glass Cleaner 1.5012531328320802
Vinegar 1.256281407035176
Floor Cleaner 1.1268781302170283
Paper Towels 1.334448160535117


In [17]:
# better way
expensive_items = []

for i in range(len(cleaningsupplies_list)):
    
    if ((country_price[i] - city_price[i])/city_price[i]) > 0.1:
        expensive_items.append(cleaningsupplies_list[i])
        
print(expensive_items)

['Dustpan', 'Garbage Bags', 'Glass Cleaner', 'Vinegar', 'Floor Cleaner', 'Paper Towels']


# Day Six: descriptive statistics
Tutorial
Part of the job of data scientists and data analysts is to understand the data and extrapolate certain findings. In most cases, the first step of understanding a numerical data set is to look at the descriptive statistics. Descriptive statistics are brief descriptions that summarize datasets. There are many great Python libraries we can use to get descriptive statistics from numerical datasets. One such library is Pandas. We won't be going too in-depth with pandas today, just the basics to understand how to use pandas to get some general descriptive statistics. We will be covering pandas extensively and going over some of the advanced functions starting on Day 8.

Pandas is one of the most widely used Python plugins. Pandas can be used when working with large data sets or performing data cleaning, manipulation, and analysis.

Since the Pandas library is external to base Python, we will need to import it. When importing, we give an alias to pandas to shorten our code when calling on functions from pandas. Instead of writing pandas.name_of_function() we'll be able to write pd.name_of_function(). The alias pd is standard within the Python community.

import pandas as pd # we must import our external python plugin

list_of_num = [1,2,3,4,5]

series = pd.Series(list_of_num) #converting our list_of_num to a pandas series variable
                                #we need to do this to use some of pandas' useful descriptive statistics functions

series.max()    #outputs maximum value in Pandas Series
series.min()    #outputs minimum value in Pandas Series
series.mean()   #outputs average value in Pandas Series
series.median() #outputs median value in Pandas Series
series.mode()   #outputs mode value in Pandas Series


### Challenge:
There are many holes in the living room's ceiling that desperately need to be fixed. Dot's measured them, and in total there are about 100. They need to figure out how much does it cost to fix all of the holes. Differently sized holes will require differently sized patches to fix them.

Size of Hole	Cost to Fix
Small (less than 20 mm)	1.30
Medium (above or equal to 20 mm AND less than 70mm)	1.60
Large (above or equal to 70 mm)	2.10
Dot needs you to look at the measurements and figure out the answers to the following questions:

What is the average sized hole?
What is the average cost to fix a hole?
What is the total cost of fixing all of the holes?

Note: Use a for loop and an if else statement to answer Q3.
Stretch Question:

Stretch Questions are not required to be completed to finish the challenge but are recommended to further develop your skills.

What is the maximum sized hole?
What is the minimum sized hole?

In [18]:
import random 
random.seed(34)

hole_sizes = [random.randint(1, i) for i in range(1, 101)]
random.shuffle(hole_sizes)

# hole sizes in mm
hole_sizes[:5]

[9, 3, 1, 21, 35]

In [19]:
print(hole_sizes)

[9, 3, 1, 21, 35, 53, 23, 14, 61, 39, 9, 17, 16, 75, 48, 37, 18, 1, 53, 38, 30, 44, 12, 14, 46, 26, 3, 9, 96, 3, 43, 44, 60, 14, 4, 68, 54, 24, 25, 29, 39, 18, 1, 8, 57, 21, 5, 69, 67, 46, 27, 37, 47, 12, 59, 16, 10, 14, 36, 2, 9, 79, 52, 28, 35, 30, 29, 5, 11, 1, 52, 67, 2, 40, 26, 4, 19, 17, 7, 58, 23, 20, 38, 18, 1, 25, 39, 26, 27, 93, 8, 12, 2, 12, 53, 9, 1, 40, 6, 5]


In [22]:
import pandas as pd
series = pd.Series(hole_sizes)

In [24]:
series.mean()

28.39

In [27]:
total = 0.0
for i in range(len(hole_sizes)):
    if hole_sizes[i] < 20:
        total += 1.3
    elif hole_sizes[i] < 70:
        total += 1.6
    else:
        total += 2.1

In [28]:
print(total)

148.79999999999987


In [29]:
round(total,2)

148.8

In [30]:
# better way
count_small = 0
count_medium = 0
count_large = 0 

for i in hole_sizes:
    if i < 20:
        count_small += 1
    elif i >= 20 and i < 70:
        count_medium += 1       
    else:
        count_large += 1
        
print(f'The amount of small holes is {count_small}.')
print(f'The amount of small holes is {count_medium}.')
print(f'The amount of small holes is {count_large}.')

The amount of small holes is 44.
The amount of small holes is 52.
The amount of small holes is 4.


In [31]:
# better way
average_cost = ((count_small*1.3)+(count_medium*1.6)+(count_large*2.1))/ 100
print(round(average_cost,4))

1.488


In [32]:
# better way
total_cost = 0

for i in hole_sizes:
    if i < 20:
        total_cost += 1.3
    elif i >= 20 and i < 70:
        total_cost += 1.6       
    else:
        total_cost += 2.1
        
print(round(total_cost,4))

148.8
